molt fetch transformation rules (#18797)
cockroachdb/docs: src/current/molt/molt-fetch.md

	- `schema` specifies the schemas to be affected by the transformation rule, formatted as a POSIX regex string.
	- `table` specifies the tables to be affected by the transformation rule, formatted as a POSIX regex string.
- `column_exclusion_opts` configures the following options for column exclusions and computed columns:
	- `column` specifies source columns to exclude from being mapped to regular columns on the target schema. It is formatted as a POSIX regex string.
	- `add_computed_def`, when set to `true`, specifies that each matching `column` should be mapped to a [computed column]({% link {{ site.current_cloud_version }}/computed-columns.md %}) on the target schema, using the computed column definition on the source schema. This assumes that all matching columns are computed columns on the source.
		{{site.data.alerts.callout_danger}}
		Columns that match the `column` regex will **not** be moved to CockroachDB if either `add_computed_def` is omitted, or a matching column is a non-computed column.
		{{site.data.alerts.end}}
- `table_rename_opts` configures the following option for table renaming:

> 	- `value` specifies the table name to which the matching `table` is mapped. If there is only one matching `table` on the source, it is renamed to `value` on the target. If there is more than one matching `table`, this assumes that all matching tables are [partitioned tables]({% link {{ site.current_cloud_version }}/partitioning.md %}) with the same schema. Otherwise, the process will error.

Reviewer: ```suggestion
	- `value` specifies the table name to which the matching `resource_specifier` is mapped. If there is only one table matching `resource_specifier` on the source, it is renamed to `table_rename_opts.value` on the target. If there is more than one matching `resource_specifier` (i.e. `n` to `1` mapping), this assumes that all matching tables are [partitioned tables]({% link {{ site.current_cloud_version }}/partitioning.md %}) with the same schema. Otherwise, the process will error. 
	- Additionally, in such `n` to `1` mapping situation: 
	     - Since these tables' data will be ingested into the single target table _concurrently_, only `--use-copy` or `--direct-copy` mode is allowed for the data ingestion phase. 
	     - User will have to create the schema of the target table on their own, and must choose `--table-handling=none` mode. I.e. `--table-handling={truncate|drop-on-target-and-recreate}` are disabled for this case.
```

The preceding JSON example therefore defines two rules:

- Rule `1` maps all source `age` columns on the source database to [computed columns]({% link {{ site.current_cloud_version }}/computed-columns.md %}) on CockroachDB. This assumes that all matching `age` columns are defined as computed columns on the source.

> - Rule `2` maps all table names beginning with `charges_part` from the source database to a single `charges` table on CockroachDB. This assumes that all matching `charges_part.*` tables have the same schema.

Reviewer: ```suggestion
- Rule `2` maps all table names with prefix `charges_part` from the source database to a single `charges` table on CockroachDB. (I.e. an `n` to `1` mapping) This assumes that all matching `charges_part.*` tables have the same schema.
```

- `column` specifies the column that will use the custom type mapping in `type-kv`. If `*` is specified, then all columns in the `table` with the matching `source-type` are converted.
- `type-kv` specifies the `source-type` that maps to the target `crdb-type`.

### Transformations

You can define transformation rules to be performed on the target schema during the fetch process. These can be used to:

- Map [computed columns]({% link {{ site.current_cloud_version }}/computed-columns.md %}) to a target schema.
- Map [partitioned tables]({% link {{ site.current_cloud_version }}/partitioning.md %}) to a single target table.
- Rename tables on the target schema.

> Transformation rules are specified using a JSON file and `--transformations-file`. For example:

Author: Took this suggestion!

~~~ json
{
  "transforms": [
    {
      "id": 1,
      "resource_specifier": {
        "schema": ".*",
        "table": ".*"
      },
      "column_exclusion_opts": {
        "add_computed_def": true,
        "column": "^age$"
      }
    },
    {
      "id": 2,
      "resource_specifier": {
        "schema": "public",
        "table": "charges_part.*"
      },
      "table_rename_opts": {
        "value": "charges"
      }
    }
  ]
}
~~~

- `resource_specifier` configures the following options for transformation rules:
	- `schema` specifies the schemas to be affected by the transformation rule, formatted as a POSIX regex string.
	- `table` specifies the tables to be affected by the transformation rule, formatted as a POSIX regex string.
- `column_exclusion_opts` configures the following options for column exclusions and computed columns:
	- `column` specifies source columns to exclude from being mapped to regular columns on the target schema. It is formatted as a POSIX regex string.

> 	- `add_computed_def`, when set to `true`, specifies that each matching `column` should be mapped to a [computed column]({% link {{ site.current_cloud_version }}/computed-columns.md %}) on the target schema. Instead of being moved from the source, the column data is generated on the target using [`ALTER TABLE ... ADD COLUMN`]({% link {{ site.current_cloud_version }}/alter-table.md %}#add-column) and the computed column definition from the source schema. This assumes that all matching columns are computed columns on the source.

Reviewer 2: Question: what kind of computed column is it mapped to `STORED` or `VIRTUAL`?

Author: I had the same question; it is the same as the source. This also comes up later in the doc and is hopefully clearer.

		{{site.data.alerts.callout_danger}}

> 		Columns that match the `column` regex will **not** be moved to CockroachDB if either `add_computed_def` is omitted, or a matching column is a non-computed column.

Reviewer 3: Might be worthwhile to add a point that if `add_computed_def` the data will not be moved over from the source, but rather the `ALTER TABLE` will be applied to add the computed column.

> 		Columns that match the `column` regex will **not** be moved to CockroachDB if either `add_computed_def` is omitted, or a matching column is a non-computed column.

Author: @ryanluu12345 Is this accurate?

> 		Columns that match the `column` regex will **not** be moved to CockroachDB if either `add_computed_def` is omitted, or a matching column is a non-computed column.

Reviewer 3: Set to false is the same as omitted. The default value is considered false when being processed.

		{{site.data.alerts.end}}
- `table_rename_opts` configures the following option for table renaming:

> 	- `value` specifies the table name to which the matching `resource_specifier` is mapped. If there is only one table matching `resource_specifier` on the source, it is renamed to `table_rename_opts.value` on the target. If there is more than one matching `resource_specifier` (i.e., an n-to-1 mapping), this assumes that all matching tables are [partitioned tables]({% link {{ site.current_cloud_version }}/partitioning.md %}) with the same schema. Otherwise, the process will error. 

Author: Took this suggestion too!

	- Additionally, in such `n` to `1` mapping situation: 
	     - Since these tables' data will be ingested into the single target table _concurrently_, only `--use-copy` or `--direct-copy` mode is allowed for the data ingestion phase. 

> 	     - User will have to create the schema of the target table on their own, and must choose `--table-handling=none` mode. I.e. `--table-handling={truncate|drop-on-target-and-recreate}` are disabled for this case.

Reviewer 2: ```suggestion
	     - User will have to create the schema of the target table on their own, and must choose [`--table-handling=none` mode](#target-table-handling). The other modes, i.e. `truncate` and `drop-on-target-and-recreate`, are disabled for this case.
```

Author: I completely missed these lines being committed. I will do another pass.

The preceding JSON example therefore defines two rules:

- Rule `1` maps all source `age` columns on the source database to [computed columns]({% link {{ site.current_cloud_version }}/computed-columns.md %}) on CockroachDB. This assumes that all matching `age` columns are defined as computed columns on the source.
- Rule `2` maps all table names with prefix `charges_part` from the source database to a single `charges` table on CockroachDB (i.e., an n-to-1 mapping). This assumes that all matching `charges_part.*` tables have the same schema.

Each rule is applied in the order it is defined. If two rules overlap, the later rule will override the earlier rule.

> To verify that the logging shoes that the computed columns are being created:

Reviewer 3: Nit: `logging shoes` should be `logging shows`.

When running `molt fetch`, set `--logging 'debug'` and look for `ALTER TABLE ... ADD COLUMN` statements with the `STORED` or `VIRTUAL` keywords in the log output:

~~~ json
{"level":"debug","time":"2024-07-22T12:01:51-04:00","message":"running: ALTER TABLE IF EXISTS public.computed ADD COLUMN computed_col INT8 NOT NULL AS ((col1 + col2)) STORED"}
~~~

> After running `molt fetch`, show the `CREATE` statement for a table:

Reviewer 3: Ensure that this statement is run on the target. For a sanity check, they can also run `SELECT * from public.computed LIMIT 1`

- `column` specifies the column that will use the custom type mapping in `type-kv`. If `*` is specified, then all columns in the `table` with the matching `source-type` are converted.
- `type-kv` specifies the `source-type` that maps to the target `crdb-type`.

### Transformations

You can define transformation rules to be performed on the target schema during the fetch process. These can be used to:

- Map [computed columns]({% link {{ site.current_cloud_version }}/computed-columns.md %}) to a target schema.
- Map [partitioned tables]({% link {{ site.current_cloud_version }}/partitioning.md %}) to a single target table.
- Rename tables on the target schema.

> Transformation rules are defined in the JSON file indicated by the `--transformations-file` flag. The JSON is formatted as follows:

Reviewer 2: The code snippet contains examples of transformation rules. The rules are formatted in JSON. 

```suggestion
Transformation rules are defined in the JSON file indicated by the `--transformations-file` flag. Two examples of rules in JSON format follow:
```
